Add semantic highlighting blog (#3858)
opensearch-project/project-website: _posts/2025-07-07-semantic-highlighting.md

**Lexical highlighter**

> To make a red <mark>pasta</mark> sauce, start by adding water, tomato paste, and diced tomatoes to a large saucepan. Then, sprinkle in some finely-grated carrots, diced onions, chopped garlic, and some spices like celery salt, dried oregano, and dried basil. Next, bring everything to a boil over medium heat before reducing the temperature to low. Finally, cover the pot and simmer the sauce for 15-30 minutes.

**Semantic highlighter**

> To make a red pasta sauce, start by adding water, tomato paste, and diced tomatoes to a large saucepan. Then, sprinkle in some finely-grated carrots, diced onions, chopped garlic, and some spices like celery salt, dried oregano, and dried basil. Next, bring everything to a boil over medium heat before reducing the temperature to low. <mark>Finally, cover the pot and simmer the sauce for 15-30 minutes.</mark>

In this example, the lexical highlighter finds a direct word match ("pasta") but misses the sentence that actually answers the query. By contrast, the semantic highlighter identifies the answer that corresponds to the intent behind the question.

## When to use semantic highlighting

> Semantic highlighting is especially useful in these scenarios:

Reviewer: @kolchfa-aws Can we make the following change:
```suggestion
Semantic highlighting is useful in a wide variety of scenarios. Some examples include:
```
These are just example use-cases, we don't want to suggest there has been any special training or evaluation for these specific tasks.

Reviewer 2: @greaa-aws Done!

The model employs a transformer-based architecture based on [Bidirectional Encoder Representations from Transformers (BERT)](https://huggingface.co/docs/transformers/en/model_doc/bert). We jointly encode both the document and query text to generate a representation for each sentence that incorporates context from both the surrounding text in the document and from the query itself. We trained the model on a diverse set of data sources, encouraging it to learn highlighting rules that apply across a wide variety of domains and use cases. We evaluated model performance primarily in terms of highlight precision and recall on *out-of-distribution* data, with the goal of selecting a highlighting model that has robust performance beyond standard training corpora.

## Performance benchmarks

We evaluated the latency and accuracy of semantic highlighting on the [MultiSpanQA dataset](https://multi-span.github.io/). The test environment was as follows.

|**OpenSearch cluster**|Version 3.1.0 deployed using `opensearch-cluster-cdk`|
|---|---|
|**Data Nodes**|3 × r6g.2xlarge (8 vCPUs, 64 GB memory each)|
|**Coordinator Nodes**|3 × c6g.xlarge (4 vCPUs, 8 GB memory each)|

> |Semantic Highlighting Model*|`opensearch-semantic-highlighter-v1` deployed remotely at AWS SageMaker Endpoint with GPU based ml.g5.xlarge (scalable 1-3 instances)|

Reviewer 3: ```suggestion
|**Semantic highlighting model**|`opensearch-semantic-highlighter-v1` deployed remotely at Amazon SageMaker endpoint with GPU-based ml.g5.xlarge (scalable 1--3 instances)|
```

> |Embedding Model|`sentence-transformers/all-MiniLM-L6-v2` deployed within OpenSearch cluster|

Reviewer 3: ```suggestion
|**Embedding model**|`sentence-transformers/all-MiniLM-L6-v2` deployed within OpenSearch cluster|
```

|**Benchmark Client**|ARM64, 16 cores, 61GB RAM|
|**Test Configuration**|10 warmup iterations, 50 test iterations, 1 shard, 0 replicas|
|**Dataset**|MultiSpanQA (1,959 documents)|
|**Document Stats**|Mean: 1,213 chars, P50: 1,111, P90: 2,050, Max: 6,689|
|**Relevant Sentences**|1,541 (9.51% of total)|

> **Note**: The benchmark used AWS SageMaker's ml.g5.xlarge GPU instances for the semantic highlighting model, which provided significant performance improvements over OpenSearch ML local deployed model. The GPU acceleration reduced P50 latency by approximately 4.5x (from 180ms to 40ms for k=1) compared to running the same model on a locally deployed model within OpenSearch. The auto-scaling configuration (1-3 instances) ensures the endpoint can handle varying workload demands while maintaining consistent performance. For the model deployment on AWS SageMaker, see the [documentation and scripts](https://github.com/opensearch-project/opensearch-py-ml/tree/main/docs/source/examples/aws_sagemaker_sentence_highlighter_model).

Reviewer 3: ```suggestion
**Note**: The benchmark used Amazon SageMaker's ml.g5.xlarge GPU instances for the semantic highlighting model, which provided significant performance improvements compared to the locally deployed OpenSearch machine learning model. The GPU acceleration reduced P50 latency by approximately 4.5x (from 180ms to 40ms for k=1) compared to running the same model on a locally deployed model within OpenSearch. The auto-scaling configuration (1--3 instances) ensures the endpoint can handle varying workload demands while maintaining consistent performance. For the model deployment on Amazon SageMaker, see the [documentation and scripts](https://github.com/opensearch-project/opensearch-py-ml/tree/main/docs/source/examples/aws_sagemaker_sentence_highlighter_model).
```

### Latency

> We measured the latency of semantic search with semantic highlighting over a range of documents retrieved (k-value) and search clients. For comparison, we included the latency of the semantic search with no highlighting. The results are presented in the following table.

Reviewer 3: ```suggestion
We measured the latency of semantic search with semantic highlighting over a range of search clients and retrieved documents (k-value). For comparison, we included the latency of the semantic search with no highlighting. The results are presented in the following table.
```

|K-value|Search clients|Semantic search only P50 latency (ms)|Semantic search with semantic highlighting P50 latency (ms)|Semantic search only P90 latency (ms)|Semantic search with semantic highlighting P90 latency (ms)|Semantic search only P100 latency (ms)|Semantic search with semantic highlighting P100 latency (ms)|
|---|---|---|---|---|---|---|---|
|1|1|21|38|23|42|24|59|
|1|4|24|37|25|45|27|78|
|1|8|24|40|26|52|28|81|
|10|1|26|180|27|199|28|237|
|10|4|25|209|27|240|29|312|
|10|8|26|267|28|323|31|407|
|20|1|24|348|25|383|25|410|
|20|4|24|401|28|449|30|530|
|20|8|26|545|28|625|32|770|
|50|1|24|806|25|861|26|954|
|50|4|25|987|26|1,074|29|1,162|
|50|8|26|1,358|28|1,490|32|1,687|

> Our comprehensive benchmarking demonstrates that the feature performs well for typical search scenarios (k≤10), providing sub-200ms responses that meet the requirements of interactive applications. Latency increases with the number of documents returned, reflecting the additional costs of inference for the semantic highlighter model.

Reviewer 3: ```suggestion
Our comprehensive benchmarking demonstrates that the feature performs well for typical search scenarios (k≤10), providing sub-200ms responses that meet the requirements of interactive applications. Latency increases with the number of documents returned, reflecting the additional costs of inference for the semantic highlighting model.
```

### Accuracy

We measured the accuracy of the highlighter by computing the precision, recall, and F1 score of the sentence-level highlights on the MultiSpanQA validation set. The results are presented in the following table.

|Metric|Value|Description|
|---|---|---|
|**Precision**|66.40%| The percentage of highlighted sentences that are actually relevant.|
|**Recall**|79.20%|The percentage of relevant sentences that were successfully highlighted.|

> |**F1 Score**|72.20%|The harmonic mean balancing precision and recall.|

Reviewer 3: ```suggestion
|**F1 score**|72.20%|The harmonic mean balancing precision and recall.|
```

> The highlighter demonstrates strong recall (79.2%) while maintaining robust precision (66.4%), resulting in a solid F1 score of 72.2%. This performance profile is well-suited for search applications where capturing most relevant content is important while keeping false positives manageable.

Reviewer 3: ```suggestion
The highlighter demonstrated strong recall (79.2%) while maintaining robust precision (66.4%), resulting in a solid F1 score of 72.2%. This performance profile is well suited for search applications where it is important to capture the most relevant content while keeping false positives manageable.
```

> In practice, the accuracy of the highlighter may vary depending on your data. We trained the highlighter model on a diverse selection of datasets to encourage good performance in many domains, but accuracy may still decrease if the data is sufficiently different from this training set.

Reviewer 3: ```suggestion
In practice, the accuracy of the highlighter may vary depending on your data. We trained the highlighting model on a diverse selection of datasets to encourage high performance in many domains, but accuracy may still decrease if your data is substantially different from this training set.
```

The OpenSearch semantic highlighting feature is designed to work with different sentence highlighting models deployed in the OpenSearch ML Commons plugin. If you have a specific domain or task, you can train and deploy your own sentence highlighting model compatible with the ML Commons framework.

If you're interested in the specifics of preparing a custom model, including model tracing and the CI processes involved, explore the [`opensearch-py-ml` GitHub repository](https://github.com/opensearch-project/opensearch-py-ml/tree/main). This repository provides tools and examples that can help guide you in bringing your own models to OpenSearch. Once your custom model is prepared and deployed, you can reference your custom `model_id` in the highlight options.

## What's next?

Semantic highlighting in OpenSearch represents a significant advancement in search result presentation. Highlighting content based on semantic relevance rather than just keyword matches provides more meaningful, context-aware search results.

> This feature offers an enhanced user experience whether you're searching through product catalogs, research papers, legal documents, or any text-based content. We invite you to try semantic highlighting and share your feedback with the OpenSearch community.

Reviewer 3: ```suggestion
This feature offers an enhanced user experience, whether you're searching through product catalogs, research papers, legal documents, or any text-based content. We invite you to try semantic highlighting and share your feedback with the OpenSearch community.
```

> We are considering several enhancements to improve semantic highlighting:

Reviewer 3: ```suggestion
We are considering several improvements to the semantic highlighting feature:
```

has_science_table: true
---

OpenSearch 3.0 introduces [semantic highlighting](https://docs.opensearch.org/docs/latest/search-plugins/searching-data/highlight/#the-semantic-highlighter), an AI-powered feature that identifies and returns the most relevant passages in a retrieved document. In this post, we'll share the motivation for this feature, explain the science behind the AI model, and show how you can incorporate semantic highlighting into your search queries.

## What is semantic highlighting?

Highlighting is a search feature that extracts the parts of a document most relevant to a query. Semantic highlighting introduces a new highlighter in OpenSearch that works differently from existing ones in two key ways: it measures relevance based on semantic similarity between the query and the text, and it highlights spans of text rather than exact keyword matches. An AI model evaluates each sentence, using context from both the query and surrounding text to determine relevance.

> This feature is designed for semantic and hybrid search use cases, where users care more about the meaning of their query than the exact words. Semantic highlighting extends that idea by surfacing the most meaningful passages within documents.

Reviewer 2: Changed to "AI search" for a comprehensive term.
